src/Uno.UI.RemoteControl.Server.Processors/HotReload/ServerHotReloadProcessor.cs

```diff
@@ -530,9 +552,41 @@ private async Task ProcessUpdateFile(UpdateFile message)
 					return (FileUpdateResult.FileNotFound, $"Requested file '{message.FilePath}' does not exists.");
 				}
 
+				var effectiveUpdate = WaitForFileUpdated(message.FilePath);
 				File.Delete(message.FilePath);
+				await effectiveUpdate;
+
 				return (FileUpdateResult.Success, null);
 			}
+
+			static async Task WaitForFileUpdated(string path)
+			{
+				var file = new FileInfo(path);
+				var dir = file.Directory;
+				while (dir is { Exists: false })
+				{
+					dir = dir.Parent;
+				}
+
+				if (dir is null)
+				{
+					return;
+				}
+
+				var tcs = new TaskCompletionSource();
+				using var watcher = new FileSystemWatcher(dir.FullName);
+				watcher.Changed += async (snd, e) =>
+				{
+					if (e.FullPath.Equals(file.FullName, StringComparison.OrdinalIgnoreCase))
+					{
+						await Task.Delay(500);
+						tcs.TrySetResult();
+					}
+				};
+				watcher.EnableRaisingEvents = true;
+
+				await Task.WhenAny(tcs.Task, Task.Delay(TimeSpan.FromSeconds(2)));
```

Comment:
Since we're putting a timeout on how long to await this task, we should wrap this in a try-catch so we can log when the timeout is exceeded - this would aid debugging if we come across scenarios where the timeout is hit and we can't trace why HR isn't behaving

Reply:
I've raised this as a follow up issue: https://github.com/unoplatform/uno/issues/18335

Reply:
For logging purpose yes it would make sense indeed. But FYI from a HR perspective, we have a second step that will re-request a HR to the IDE if the delay is too short. And only then the client might get an invalid "no changes".

About exception handling and possible crash due to an handled exception : this not really a "timeout", it's only a task that will complete at some point, it won't raise any exception (especially since there is no `CancellationToken` here). The only thing that can throw is the creation of the task itself (e.g. configuration of the `watcher`) which is already catch by the server.